### PR TITLE
Correctly manage Content-Length on HEAD responses

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation

When we receive a HEAD response, it's possible that the response contains a content-length. llhttp has a bug
(https://github.com/nodejs/llhttp/issues/202) that prevents it from properly managing that issue, which causes us to incorrectly parse responses.

Modifications

Forcibly set llhttp's content-length value to 0.

Result

Correctly handle HTTP framing around llhttp's issues.

### CODE REVIEW

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 1]
```diff
@@ -113,6 +113,9 @@ private class BetterHTTPParser {
         }
         self.settings.pointee.on_message_complete = { opaque in
             BetterHTTPParser.fromOpaque(opaque).didReceiveMessageCompleteNotification()
+            // Temporary workaround for https://github.com/nodejs/llhttp/issues/202, should be removed
+            // when that issue is fixed.
+            opaque?.pointee.content_length = 0
```
Comment (FranzBusch) [inline Sources/NIOHTTP1/HTTPDecoder.swift:118]: Can we track this in one of our GH issues as well?

**Review by FranzBusch — APPROVED** [review]
LGTM!

File: Sources/NIOHTTP1/HTTPDecoder.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPDecoder.swift:118]: Can we repurpose #2274 to track this?

Comment (FranzBusch) [reply Sources/NIOHTTP1/HTTPDecoder.swift]: We sure can!

**Review by dnadoba — APPROVED** [review]

**Comment by fabianfett** [comment]
@Lukasa Should we add this to the workarounds.md?

**Comment by Lukasa** [comment]
Yes. I was hoping that upstream would turn around a fix quickly, but it seems like they're not in a rush to do so.

**Comment by Lukasa** [comment]
Just shipped as 2.43.0.

**Comment by adam-fowler** [comment]
Sorry to be annoying but can we get this in a v2.42.1. Builds dependent on swift 5.4 are currently broken as the fix is not available for 5.4. I know you've moved on from supporting 5.4 but it would be good to leave the build in a working state (as much as you can know about).

**Comment by Lukasa** [comment]
How many builds are dependent on 5.4? Swift 5.4 is by this point 18 months old and I'd have expected users to move forward to at least 5.5 by now.

I'm not _strongly_ opposed to back porting this fix, but in general we are not going to continue to backport every bugfix to 5.4. That's what it means for us to have dropped support.

**Comment by Lukasa** [comment]
See #2289.

**Comment by adam-fowler** [comment]
I understand that, but this was a fairly sizeable component that was replaced just before you dropped a Swift version.
Regarding number of builds dependent on 5.4 I don't know.

**Comment by adam-fowler** [comment]
thanks

**Comment by Lukasa** [comment]
Shipped as 2.42.1.

**Pull request merged**